File: iteration2/CourseRegistrationSystem/src/CourseRegistrationSystem.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.sun.org.apache.xml.internal.security.keys.storage.implementations.CertsInFilesystemDirectoryResolver;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class CourseRegistrationSystem {
    private Student student;
    private List<Student> students;
    private List<Course> courses;

    public CourseRegistrationSystem(Student student, List<Course> courses){
        this.student = student;
        this.courses = courses;
    }
    public CourseRegistrationSystem(){

    }
    public List<Course>listAvailableCourses(){
        List<Course> availableCourse = new ArrayList<>();
        List<Course> coursesTaken= new ArrayList<>();
        for (Grade gr:student.viewTranscript().allGrades()){
            coursesTaken.add(gr.getCourse());
        }
        for(Course course : courses){

            if(!(student.getEnrolledCourses().stream().anyMatch(p -> p.getCourseId().equals(course.getCourseId()))) &&
                    !(coursesTaken.stream().anyMatch(p -> p.getCourseId().equals(course.getCourseId())))) {
                if (course.hasPrerequisite()){
                    for (Course crs : coursesTaken){
                        if (crs.getCourseId().equals(course.getPrerequisiteLessonId())){
                            availableCourse.add(course);
                        }
                    }
                } else{
                    if (student.getGpa()>=3){
                        availableCourse.add(course);
                    } else if (student.getSemester()>=course.getCourseSection().getSemester()){
                        availableCourse.add(course);
                    }

                }
            }
        }
        return availableCourse;
    }
    public void requestInCourse(Course course, Student student) throws IOException {
        student.getRequestedCourses().add(course);

        JSONMethods jM = new JSONMethods();
        jM.addRequestCourse(course,student);
        jM.updateEnrollmentCapacity(course.getCourseId(),course.getCourseSection().getEnrollmentCapacity()-1);
    }

    public void rejectCourse(Course course) throws IOException {
        JSONMethods jM = new JSONMethods();
        jM.updateEnrollmentCapacity(course.getCourseId(),course.getCourseSection().getEnrollmentCapacity()+1);

    }

    public boolean checkForConflicts(List<Course> courses) {
        System.out.println("\nChecking for schedule conflicts:");

        for (int i = 0; i < courses.size() - 1; i++) {
            Course course1 = courses.get(i);
            CourseSection section1 = course1.getCourseSection();

            for (int j = i + 1; j < courses.size(); j++) {
                Course course2 = courses.get(j);
                CourseSection section2 = course2.getCourseSection();

                if (section1.getDay().equals(section2.getDay()) &&
                        doTimeRangesOverlap(section1.getHour(), section2.getHour())) {
                    System.out.println("Conflict found between:");
                    System.out.println(course1);
                    System.out.println(course2);
                    System.out.println("Please select only one of them");
                    return true;
                }
            }
        }
        return false;
    }

    private boolean doTimeRangesOverlap(String timeRange1, String timeRange2) {
        String[] parts1 = timeRange1.split("-");
        String[] parts2 = timeRange2.split("-");

        String startHour = (parts1[0].split(":"))[0];
        String endHour = (parts1[1].split(":"))[0];

        String startHour2 = (parts2[0].split(":"))[0];


        return (Integer.parseInt(startHour2)>=Integer.parseInt(startHour) && Integer.parseInt(startHour2)<=Integer.parseInt(endHour));
    }

}

File: iteration3/CourseRegistrationSystem.py
from JsonMethods import JSONMethods


class CourseRegistrationSystem:
    def __init__(self, student, courses):
        self.student = student
        self.courses = courses

    def list_available_courses(self):
        available_courses = []
        courses_taken = [grade.get_course() for grade in self.student.view_transcript().all_grades()]
        
        for course in self.courses:
            enrolled_courses_ids = [enrolled_course.get_course_id() for enrolled_course in self.student.get_enrolled_courses()]
            
            if (course.get_course_id() not in enrolled_courses_ids and 
                course.get_course_id() not in [taken_course.get_course_id() for taken_course in courses_taken]):
                
                if course.has_prerequisite():
                    for taken_course in courses_taken:
                        if taken_course.get_course_id() == course.get_prerequisite_lesson_id():
                            available_courses.append(course)
                            break
                else:
                    if self.student.get_gpa() >= 3 or self.student.get_semester() >= course.get_course_section().get_semester():
                        available_courses.append(course)
        
        return available_courses

    def request_in_course(self, course, student):
        student.add_requested_course(course)
        jM = JSONMethods()
        jM.add_request_course(course, student)
        jM.update_enrollment_capacity(course.get_course_id(), course.get_course_section().get_enrollment_capacity() - 1)

    def reject_course(self, course):
        jM = JSONMethods()
        jM.update_enrollment_capacity(course.get_course_id(), course.get_course_section().get_enrollment_capacity() + 1)

    def check_for_conflicts(self, courses):
        for i in range(len(courses) - 1):
            course1 = courses[i]
            section1 = course1.get_course_section()

            for j in range(i + 1, len(courses)):
                course2 = courses[j]
                section2 = course2.get_course_section()

                if section1.get_day() == section2.get_day() and self.do_time_ranges_overlap(section1.get_hour(), section2.get_hour()):
                    return f"Conflict found between:\n{course1}\n{course2}\nPlease select only one of them"

        return ""

    def do_time_ranges_overlap(self, time_range1, time_range2):
        start_hour1, end_hour1 = map(int, time_range1.split("-"))
        start_hour2, end_hour2 = map(int, time_range2.split("-"))

        return start_hour1 <= start_hour2 <= end_hour1 or start_hour1 <= end_hour2 <= end_hour1
